File: Python/Petshop.py
class Petshop:
    def __init__(this):
        this.id = []
        this.nama_produk = []
        this.kategori_produk = []
        this.harga_produk = []
        this.jumlah_produk = 0
    
    def __del__(this):
        pass

    # Getter & Setter
    def get_id(this, index):
        return this.id[index]
    
    def set_id(this, index, value):
        this.id[index] = value
    
    def get_nama_produk(this, index):
        return this.nama_produk[index]
    
    def set_nama_produk(this, index, value):
        this.nama_produk[index] = value
    
    def get_kategori_produk(this, index):
        return this.kategori_produk[index]
    
    def set_kategori_produk(this, index, value):
        this.kategori_produk[index] = value
    
    def get_harga_produk(this, index):
        return this.harga_produk[index]
    
    def set_harga_produk(this, index, value):
        this.harga_produk[index] = value
    
    def get_jumlah_produk(this):
        return this.jumlah_produk
    
    def longest_id(this):
        return max((len(str(i)) for i in this.id), default=3)
    
    def longest_nama(this):
        return max((len(nama) for nama in this.nama_produk), default=4)
    
    def longest_kategori(this):
        return max((len(kategori) for kategori in this.kategori_produk), default=8)
    
    def longest_harga(this):
        return max((len(str(harga)) for harga in this.harga_produk), default=5)
    
    def tampilkan_produk(this):
        if this.jumlah_produk == 0:
            print("Data kosong, tidak ada produk di PetShop!")
            return
        
        id_width = max(3, this.longest_id())
        nama_width = max(4, this.longest_nama())
        kategori_width = max(8, this.longest_kategori())
        harga_width = max(5, this.longest_harga())
        
        print("+" + "-" * (id_width + 2) + "+" + "-" * (nama_width + 2) + "+" + "-" * (kategori_width + 2) + "+" + "-" * (harga_width + 2) + "+")
        print(f"| {'ID':<{id_width}} | {'Nama':<{nama_width}} | {'Kategori':<{kategori_width}} | {'Harga':<{harga_width}} |")
        print("+" + "-" * (id_width + 2) + "+" + "-" * (nama_width + 2) + "+" + "-" * (kategori_width + 2) + "+" + "-" * (harga_width + 2) + "+")
        
        for i in range(this.jumlah_produk):
            print(f"| {this.id[i]:<{id_width}} | {this.nama_produk[i]:<{nama_width}} | {this.kategori_produk[i]:<{kategori_width}} | {this.harga_produk[i]:<{harga_width}} |")
        
        print("+" + "-" * (id_width + 2) + "+" + "-" * (nama_width + 2) + "+" + "-" * (kategori_width + 2) + "+" + "-" * (harga_width + 2) + "+")
    
    def tambah_produk(this, temp_id, temp_nama, temp_kategori, temp_harga):
        if this.jumlah_produk >= 100:
            print("PetShop sudah penuh, tidak bisa menambah produk lagi!")
            return
        
        if temp_id in this.id:
            print(f"Produk dengan ID {temp_id} sudah ada! Gagal menambahkan produk.")
            return
        
        this.id.append(temp_id)
        this.nama_produk.append(temp_nama)
        this.kategori_produk.append(temp_kategori)
        this.harga_produk.append(temp_harga)
        this.jumlah_produk += 1
        print("Produk berhasil ditambahkan!")
    
    def edit_produk(this, nama, temp_id, temp_nama, temp_kategori, temp_harga):
        if this.jumlah_produk == 0:
            print("Tidak ada produk untuk diedit!")
            return
        
        for i in range(this.jumlah_produk):
            if this.nama_produk[i] == nama:
                this.id[i] = temp_id
                this.nama_produk[i] = temp_nama
                this.kategori_produk[i] = temp_kategori
                this.harga_produk[i] = temp_harga
                print("Produk berhasil diperbarui!")
                return
        print("Produk tidak ditemukan!")
    
    def hapus_produk(this, nama):
        if this.jumlah_produk == 0:
            print("Tidak ada produk untuk dihapus!")
            return
        
        for i in range(this.jumlah_produk):
            if this.nama_produk[i] == nama:
                del this.id[i]
                del this.nama_produk[i]
                del this.kategori_produk[i]
                del this.harga_produk[i]
                this.jumlah_produk -= 1
                print("Produk berhasil dihapus!")
                return
        print("Produk tidak ditemukan!")
    
    def cari_produk(this, nama):
        if this.jumlah_produk == 0:
            print("Tidak ada produk untuk dicari!")
            return
        
        for i in range(this.jumlah_produk):
            if this.nama_produk[i] == nama:
                print(f"Produk ditemukan!\nID: {this.id[i]}, Kategori: {this.kategori_produk[i]}, Harga: {this.harga_produk[i]}")
                return
        print("Produk tidak ditemukan!")

File: C++/Petshop.cpp
#include <iostream>
#include <iomanip>
using namespace std;

const int MAX_PRODUK = 100;

class PetShop
{
private:
    int id[MAX_PRODUK];
    string namaProduk[MAX_PRODUK];
    string kategoriProduk[MAX_PRODUK];
    int hargaProduk[MAX_PRODUK];
    int jumlahProduk;

public:
    PetShop()
    {
        jumlahProduk = 0;
    }
    
    ~PetShop()
    {
    }

    // Getter & Setter
    int getId(int index)
    {
        return id[index];
    }
    void setId(int index, int value)
    {
        id[index] = value;
    }

    string getNamaProduk(int index)
    {
        return namaProduk[index];
    }
    void setNamaProduk(int index, string value)
    {
        namaProduk[index] = value;
    }

    string getKategoriProduk(int index)
    {
        return kategoriProduk[index];
    }
    void setKategoriProduk(int index, string value)
    {
        kategoriProduk[index] = value;
    }

    int getHargaProduk(int index)
    {
        return hargaProduk[index];
    }
    void setHargaProduk(int index, int value)
    {
        hargaProduk[index] = value;
    }

    int getJumlahProduk()
    {
        return jumlahProduk;
    }

    void idDigit(int id)
    {
        if (id < 10) // 1 digit
        {
            cout << "00" << id;
        }
        else if (id < 100) // 2 digit
        {
            cout << "0" << id;
        }
        else
        {
            cout << id;
        }
    }

    int longestID()
    {
        int longest = 0;
        for (int i = 0; i < jumlahProduk; i++)
        {
            if (to_string(id[i]).length() > longest)
            {
                longest = to_string(id[i]).length();
            }
        }
        return longest;
    }

    int longestNama()
    {
        int longest = 0;
        for (int i = 0; i < jumlahProduk; i++)
        {
            if (namaProduk[i].length() > longest)
            {
                longest = namaProduk[i].length();
            }
        }
        return longest;
    }

    int longestKategori()
    {
        int longest = 0;
        for (int i = 0; i < jumlahProduk; i++)
        {
            if (kategoriProduk[i].length() > longest)
            {
                longest = kategoriProduk[i].length();
            }
        }
        return longest;
    }

    int longestHarga()
    {
        int longest = 0;
        for (int i = 0; i < jumlahProduk; i++)
        {
            if (to_string(hargaProduk[i]).length() > longest)
            {
                longest = to_string(hargaProduk[i]).length();
            }
        }
        return longest;
    }

    // CRUD
    void tampilkanProduk()
    {
        if (jumlahProduk == 0)
        {
            cout << "Data kosong, tidak ada produk di PetShop!\n";
            return;
        }

        // Mencari lebar kolom maksimum
        int idWidth = max(3, longestID());           
        int namaWidth = max(4, longestNama());        
        int kategoriWidth = max(8, longestKategori()); 
        int hargaWidth = max(5, longestHarga());       

        // Mencetak border atas tabel
        cout << "+" << string(idWidth + 2, '-') << "+"
             << string(namaWidth + 2, '-') << "+"
             << string(kategoriWidth + 2, '-') << "+"
             << string(hargaWidth + 2, '-') << "+\n";

        // Mencetak header
        cout << "| " << left << setw(idWidth) << "ID"
             << " | " << setw(namaWidth) << "Nama"
             << " | " << setw(kategoriWidth) << "Kategori"
             << " | " << setw(hargaWidth) << "Harga"
             << " |\n";

        // Mencetak border bawah header
        cout << "+" << string(idWidth + 2, '-') << "+"
             << string(namaWidth + 2, '-') << "+"
             << string(kategoriWidth + 2, '-') << "+"
             << string(hargaWidth + 2, '-') << "+\n";

        // Mencetak data produk
        for (int i = 0; i < jumlahProduk; i++)
        {
            cout << "| " << setw(idWidth) << id[i]
                 << " | " << setw(namaWidth) << namaProduk[i]
                 << " | " << setw(kategoriWidth) << kategoriProduk[i]
                 << " | " << setw(hargaWidth) << hargaProduk[i]
                 << " |\n";
        }

        // Mencetak border bawah tabel
        cout << "+" << string(idWidth + 2, '-') << "+"
             << string(namaWidth + 2, '-') << "+"
             << string(kategoriWidth + 2, '-') << "+"
             << string(hargaWidth + 2, '-') << "+\n";
    }

    void tambahProduk()
    {
        if (jumlahProduk >= MAX_PRODUK)
        {
            cout << "PetShop sudah penuh, tidak bisa menambah produk lagi!\n";
            return;
        }

        int tempId, tempHarga;
        string tempNama, tempKategori;

        cout << "Masukkan ID: ";
        cin >> tempId;
        cin.ignore();

        // Cek apakah ID sudah ada
        for (int i = 0; i < jumlahProduk; i++)
        {
            if (getId(i) == tempId)
            {
                cout << "Produk dengan ID " << tempId << " sudah ada! Gagal menambahkan produk.\n";
                return;
            }
        }

        cout << "Masukkan Nama Produk: ";
        getline(cin, tempNama);

        cout << "Masukkan Kategori Produk: ";
        getline(cin, tempKategori);

        cout << "Masukkan Harga Produk: ";
        cin >> tempHarga;

        setId(jumlahProduk, tempId);
        setNamaProduk(jumlahProduk, tempNama);
        setKategoriProduk(jumlahProduk, tempKategori);
        setHargaProduk(jumlahProduk, tempHarga);

        jumlahProduk++;
        cout << "Produk berhasil ditambahkan!\n";
    }

    void editProduk()
    {
        if (jumlahProduk == 0)
        {
            cout << "Tidak ada produk untuk diedit!\n";
            return;
        }

        string nama;
        cin.ignore();
        cout << "Masukkan nama produk yang ingin diedit: ";
        getline(cin, nama);

        for (int i = 0; i < jumlahProduk; i++)
        {
            if (getNamaProduk(i) == nama)
            {
                int tempId, tempHarga;
                string tempNama, tempKategori;

                cout << "Masukkan ID baru: ";
                cin >> tempId;
                cin.ignore();

                cout << "Masukkan Nama Produk baru: ";
                getline(cin, tempNama);

                cout << "Masukkan Kategori Produk baru: ";
                getline(cin, tempKategori);

                cout << "Masukkan Harga Produk baru: ";
                cin >> tempHarga;

                setId(i, tempId);
                setNamaProduk(i, tempNama);
                setKategoriProduk(i, tempKategori);
                setHargaProduk(i, tempHarga);

                cout << "Produk berhasil diperbarui!\n";
                return;
            }
        }
        cout << "Produk tidak ditemukan!\n";
    }

    void hapusProduk()
    {
        if (jumlahProduk == 0)
        {
            cout << "Tidak ada produk untuk dihapus!\n";
            return;
        }

        string nama;
        cin.ignore();
        cout << "Masukkan nama produk yang ingin dihapus: ";
        getline(cin, nama);

        for (int i = 0; i < jumlahProduk; i++)
        {
            if (getNamaProduk(i) == nama)
            {
                for (int j = i; j < jumlahProduk - 1; j++)
                {
                    setId(j, getId(j + 1));
                    setNamaProduk(j, getNamaProduk(j + 1));
                    setKategoriProduk(j, getKategoriProduk(j + 1));
                    setHargaProduk(j, getHargaProduk(j + 1));
                }
                jumlahProduk--;
                cout << "Produk berhasil dihapus!\n";
                return;
            }
        }
        cout << "Produk tidak ditemukan!\n";
    }

    void cariProduk()
    {
        if (jumlahProduk == 0)
        {
            cout << "Tidak ada produk untuk dicari!\n";
            return;
        }

        string nama;
        cin.ignore();
        cout << "Masukkan nama produk yang dicari: ";
        getline(cin, nama);

        for (int i = 0; i < jumlahProduk; i++)
        {
            if (getNamaProduk(i) == nama)
            {
                cout << "Produk ditemukan!\n"
                     << "ID: " << getId(i)
                     << ", Kategori: " << getKategoriProduk(i)
                     << ", Harga: " << getHargaProduk(i) << endl;
                return;
            }
        }
        cout << "Produk tidak ditemukan!\n";
    }
};
